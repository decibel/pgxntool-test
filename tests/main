#!/bin/bash

trap 'echo "ERROR: $BASH_SOURCE: line $LINENO" >&2' ERR
set -o errexit -o errtrace -o pipefail
#set -o xtrace -o verbose

if [ "$1" == "-v" ]; then
  verboseout=1
  shift
fi
BASEDIR=`cd ${0%/*}; pwd`

. $BASEDIR/../.env
. $TOPDIR/lib.sh

DISTRIBUTION_NAME=distribution_test
EXTENSION_NAME=pgxntool-test # TODO: rename to something less likely to conflict


#./clone.sh "$PGXNBRANCH" "$PGXNREPO"

cd $TEST_REPO

out Making checkout dirty
touch garbage
git add garbage
out Verify setup.sh errors out
if pgxntool/setup.sh; then
  echo "setup.sh should have exited non-zero" >&2
  exit 1
fi
git reset HEAD garbage
rm garbage

out Running setup.sh
pgxntool/setup.sh

out -v Status
ls
git status
git diff


out git commit
git commit -am "Test setup"


out Verify changing META.in.json works
# Need to sleep 1 second otherwise make won't pickup new timestamp
sleep 1
# Sanity check
[ -n "$DISTRIBUTION_NAME" ] && [ -n "$EXTENSION_NAME" ]

sed -i '' -e "s/DISTRIBUTION_NAME/$DISTRIBUTION_NAME/" -e "s/EXTENSION_NAME/$EXTENSION_NAME/" META.in.json
#git diff -u

out -v This make will produce a bogus '"already up to date"' message for some reason
make
git commit -am "Change META"


# Note: It's easier to do this now than when the checkout is all cluttered
out Test creating a release
make dist
unzip -l ../$DISTRIBUTION_NAME-0.1.0.zip | grep .asc | awk '{print $4}'
# grep exits with 1 if it can't find anything
out Making sure ONLY TEST_DOC.asc is in the distribution
unzip -l ../$DISTRIBUTION_NAME-0.1.0.zip | grep TEST_DOC.asc | awk '{print $4}'
[ `unzip -l ../$DISTRIBUTION_NAME-0.1.0.zip | grep -c .asc` -eq 1 ] || exit 1

out "Run setup.sh again to verify it doesn't over-write things"
pgxntool/setup.sh
git diff --exit-code

out Copy stuff from template to where it normally lives
cp -R t/* .

out Add extension to deps.sql
quote='"'
sed -i '' -e "s/CREATE EXTENSION \.\.\..*/CREATE EXTENSION ${quote}$EXTENSION_NAME${quote};/" test/deps.sql

out Make certain test/output gets created
[ ! -e $TEST_REPO/test/output ] || (out "ERROR! test/output directory should not exist!"; exit 1)
make test # TODO: ensure this exits non-zero
[ -e $TEST_REPO/test/output ] || (out "ERROR! test/output directory does not exist!"; exit 1)
[ -d $TEST_REPO/test/output ] || (out "ERROR! test/output is not a directory!"; exit 1)

out And copy expected output file to output dir that should now exist
cp $TOPDIR/pgxntool-test.source test/output

out Run make test again
make test || exit 1
out -v ^^^ Should be clean output ^^^

out Remove input and output directories, make sure output is not recreated
rm -rf $TEST_REPO/test/input $TEST_REPO/test/output
make test
[ ! -e $TEST_REPO/test/output ] || (out "ERROR! test/output directory exists!"; exit 1)

# Mess with output to test make results
echo >> $TEST_REPO/test/expected/pgxntool-test.out

out Test make results
make test
out -v ^^^ Should have a diff ^^^
make results
make test
out -v ^^^ Should be clean output, BUT NOTE THERE WILL BE A FAILURE DIRECTLY ABOVE! ^^^

check_log

# vi: expandtab sw=2 ts=2
