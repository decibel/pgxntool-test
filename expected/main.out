# Verify changing META.in.json works
######################################
# This make will produce a bogus "already up to date" message for some reason
######################################
make[1]: `META.json' is up to date.
@GIT COMMIT@ Change META
 2 files changed, 4 insertions(+), 4 deletions(-)
# Test creating a release
git branch 0.1.0
git push --set-upstream origin 0.1.0
To ../fake_repo
 * [new branch]      0.1.0 -> 0.1.0
Branch 0.1.0 set up to track remote branch 0.1.0 from origin.
git archive --prefix=distribution_test-0.1.0/ -o ../distribution_test-0.1.0.zip 0.1.0
distribution_test-0.1.0/t/TEST_DOC.asc
# Making sure ONLY TEST_DOC.asc is in the distribution
distribution_test-0.1.0/t/TEST_DOC.asc
# Run setup.sh again to verify it doesn't over-write things
.gitignore already exists
META.in.json already exists
Makefile already exists
make[1]: `META.json' is up to date.
deps.sql already exists
On branch master
Your branch is ahead of 'origin/master' by 4 commits.
  (use "git push" to publish your local commits)
nothing to commit, working tree clean
If you won't be creating C code then you can:

rmdir src

If everything looks good then

git commit -am 'Add pgxntool (https://github.com/decibel/pgxntool/tree/release)'
# Copy stuff from template to where it normally lives
# Add extension to deps.sql
# Make certain test/output gets created
rm -rf ../distribution_test-0.1.0.zip sql/pgxntool-test--0.1.0.sql
rm -rf results/ regression.diffs regression.out tmp_check/ log/
cp sql/pgxntool-test.sql sql/pgxntool-test--0.1.0.sql
@INSTALL@ -c -d '@PG_LOCATION@/share/extension'
@INSTALL@ -c -m 644 ./sql/pgxntool-test--0.1.0.sql ./sql/pgxntool-test--0.1.0--0.1.1.sql ./pgxntool-test.control  '@PG_LOCATION@/share/extension/'
INVOCATION OF pg_regress
(using postmaster on Unix socket, default port)
============== dropping database "contrib_regression" ==============
DROP DATABASE
============== creating database "contrib_regression" ==============
CREATE DATABASE
ALTER DATABASE
============== installing plpgsql                     ==============
CREATE LANGUAGE
============== running regression test queries        ==============
test pgxntool-test            ... FAILED

======================
 1 of 1 tests failed. 
======================

The differences that caused some tests to fail can be viewed in the
file "@TEST_DIR@/repo/test/regression.diffs".  A copy of the test summary that you see
above is saved in the file "@TEST_DIR@/repo/test/regression.out".

make[1]: [installcheck] Error 1 (ignored)
*** @TEST_DIR@/repo/test/expected/pgxntool-test.out
--- @TEST_DIR@/repo/test/results/pgxntool-test.out
***************
*** 0 ****
--- 1,59 ----
+ \i @TEST_DIR@/repo/test/pgxntool/setup.sql
+ \i test/pgxntool/psql.sql
+ -- No status messages
+ \set QUIET true
+ -- Verbose error messages
+ \set VERBOSITY verbose
+ -- Revert all changes on failure.
+ \set ON_ERROR_ROLLBACK 1
+ \set ON_ERROR_STOP true
+ BEGIN;
+ \i test/pgxntool/tap_setup.sql
+ \i test/pgxntool/psql.sql
+ -- No status messages
+ \set QUIET true
+ -- Verbose error messages
+ \set VERBOSITY verbose
+ -- Revert all changes on failure.
+ \set ON_ERROR_ROLLBACK 1
+ \set ON_ERROR_STOP true
+ SET client_min_messages = WARNING;
+ DO $$
+ BEGIN
+ IF NOT EXISTS(SELECT 1 FROM pg_namespace WHERE nspname='tap') THEN
+   CREATE SCHEMA tap;
+ END IF;
+ END$$;
+ SET search_path = tap, public;
+ CREATE EXTENSION IF NOT EXISTS pgtap SCHEMA tap;
+ SET client_min_messages = NOTICE;
+ \pset format unaligned
+ \pset tuples_only true
+ \pset pager
+ -- vi: expandtab ts=2 sw=2
+ \i test/deps.sql
+ -- IF NOT EXISTS will emit NOTICEs, which is annoying
+ SET client_min_messages = WARNING;
+ -- Add any test dependency statements here
+ -- Note: pgTap is loaded by setup.sql
+ --CREATE EXTENSION IF NOT EXISTS ...;
+ /*
+  * Now load our extension. We don't use IF NOT EXISTs here because we want an
+  * error if the extension is already loaded (because we want to ensure we're
+  * getting the very latest version).
+  */
+ CREATE EXTENSION "pgxntool-test";
+ -- Re-enable notices
+ SET client_min_messages = NOTICE;
+ SELECT plan(1);
+ 1..1
+ SELECT is(
+   "pgxntool-test"(1,2)
+   , 3
+ );
+ ok 1
+ \i @TEST_DIR@/repo/test/pgxntool/finish.sql
+ SELECT finish();
+ \echo # TRANSACTION INTENTIONALLY LEFT OPEN!
+ # TRANSACTION INTENTIONALLY LEFT OPEN!
+ -- vi: expandtab ts=2 sw=2

======================================================================

# And copy expected output file to output dir that should now exist
# Run make test again
rm -rf  sql/pgxntool-test--0.1.0.sql
rm -rf results/ regression.diffs regression.out tmp_check/ log/
cp sql/pgxntool-test.sql sql/pgxntool-test--0.1.0.sql
@INSTALL@ -c -d '@PG_LOCATION@/share/extension'
@INSTALL@ -c -m 644 ./sql/pgxntool-test--0.1.0.sql ./sql/pgxntool-test--0.1.0--0.1.1.sql ./pgxntool-test.control  '@PG_LOCATION@/share/extension/'
INVOCATION OF pg_regress
(using postmaster on Unix socket, default port)
============== dropping database "contrib_regression" ==============
DROP DATABASE
============== creating database "contrib_regression" ==============
CREATE DATABASE
ALTER DATABASE
============== installing plpgsql                     ==============
CREATE LANGUAGE
============== running regression test queries        ==============
test pgxntool-test            ... ok

=====================
 All 1 tests passed. 
=====================

######################################
# ^^^ Should be clean output ^^^
######################################
# Remove input and output directories, make sure output is not recreated
rm -rf  sql/pgxntool-test--0.1.0.sql
rm -rf results/ regression.diffs regression.out tmp_check/ log/
cp sql/pgxntool-test.sql sql/pgxntool-test--0.1.0.sql
@INSTALL@ -c -d '@PG_LOCATION@/share/extension'
@INSTALL@ -c -m 644 ./sql/pgxntool-test--0.1.0.sql ./sql/pgxntool-test--0.1.0--0.1.1.sql ./pgxntool-test.control  '@PG_LOCATION@/share/extension/'
INVOCATION OF pg_regress
(using postmaster on Unix socket, default port)
============== dropping database "contrib_regression" ==============
DROP DATABASE
============== creating database "contrib_regression" ==============
CREATE DATABASE
ALTER DATABASE
============== installing plpgsql                     ==============
CREATE LANGUAGE
============== running regression test queries        ==============
test pgxntool-test            ... ok

=====================
 All 1 tests passed. 
=====================

